app/models/world_location_email_signup.rb

```diff
@@ -0,0 +1,80 @@
+class WorldLocationEmailSignup
+  attr_reader :feed
+
+  def initialize(feed)
+    @feed = feed
+  end
+
+  def save
+    return false unless valid?
+
+    ensure_subscriber_list_exists
+    true
+  end
+
+  def signup_url
+    ensure_subscriber_list_exists["subscriber_list"]["subscription_url"]
+  end
+
+  def topic_id
+    ensure_subscriber_list_exists["subscriber_list"]["gov_delivery_id"]
+  end
+
+  def name
+    world_location.name
+  end
+
+  def valid?
+    uri && recognised_url? && world_feed_type? && world_location.present?
```

Comment:
Could we do more logic in this method or something to reduce the spread of this across so many methods. I know you didn't write it originally.

thinking it could be like:

```
def valid?
  return false unless uri
  return false unless uri.host == Whitehall.public_host && uri.scheme == Whitehall.public_protocol

  recognised_path = Rails.application.routes.recognize_path(uri.path)
  valid_extension = File.extname(uri.path) == ".atom"   
  world_feed_type = uri.path.split("/")[1] == "world"

  recognised_path && valid_extension && world_feed_type && world_location.present?
rescue ActionController::RoutingError
  false
end
```

I'm also not sure we need this much validation? a lot of it feels superfluous. We could probably replace most of this logic with say:

```
def valid?
  uri && world_location_slug && world_location.present?
end

def world_location_slug
  uri.path.match(%r{^/world/(.*)\.atom$})[1]
end

def world_location
  @world_location ||= WorldLocation.find_by(slug: world_location_slug)
end
```

Reply:
Good shout, I didn't know how much of the validating I should pull across so I've gone with your suggestion.